src/typechecker.erl

```diff
@@ -1968,44 +2013,151 @@ do_type_check_expr_in(Env, ResTy, {'try', _, Block, CaseCs, CatchCs, AfterBlock}
 
 
 type_check_arith_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
-    case subtype(ResTy, {type, erl_anno:new(0), 'number', []}, Env#env.tenv) of
-	{true, Cs} ->
-	  {VarBinds1, Cs1} = type_check_expr_in(Env, ResTy, Arg1),
-	  {VarBinds2, Cs2} = type_check_expr_in(Env, ResTy, Arg2),
-	  {union_var_binds([VarBinds1, VarBinds2])
-	  ,constraints:combine([Cs, Cs1, Cs2])};
-	false ->
-	  throw({type_error, arith_error, Op, P, ResTy})
-    end.
+    type_check_arith_op_in(Env, number, ResTy, Op, P, Arg1, Arg2).
+
 type_check_int_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
-    case subtype(ResTy, {type, erl_anno:new(0), 'integer', []}, Env#env.tenv) of
-	{true, Cs} ->
-	  {VarBinds1, Cs1} = type_check_expr_in(Env, ResTy, Arg1),
-	  {VarBinds2, Cs2} = type_check_expr_in(Env, ResTy, Arg2),
-	  {union_var_binds([VarBinds1, VarBinds2])
-	  ,constraints:combine([Cs, Cs1, Cs2])};
-	false ->
-	  throw({type_error, int_error, Op, P, ResTy})
+    type_check_arith_op_in(Env, integer, ResTy, Op, P, Arg1, Arg2).
+
+type_check_arith_op_in(Env, Kind, ResTy, Op, P, Arg1, Arg2) ->
+    ResTy1 = glb(type(Kind), ResTy, Env#env.tenv),
+    IsNone =
+        case ResTy1 of
+            %% TODO: allow none() if checking against none()? Not clear that
+            %% this is sensible, since in that case you'd like to only require
+            %% *one* of the arguments to be none(), not both.
+            {type, _, none, []} -> true;
+            _ -> false
+        end,
+    case IsNone of
+        false ->
+            case arith_op_arg_types(Op, ResTy1) of
+                {ArgTy1, ArgTy2} ->
+                    {VarBinds1, Cs1} = type_check_expr_in(Env, ArgTy1, Arg1),
+                    {VarBinds2, Cs2} = type_check_expr_in(Env, ArgTy2, Arg2),
+                    {union_var_binds([VarBinds1, VarBinds2]),
+                     constraints:combine([Cs1, Cs2])};
+                false ->
+                    throw({type_error, arith_type_too_precise, Op, P, ResTy1})
+            end;
+        true ->
+            Tag = if Kind == integer -> int_error;
+                     true            -> arith_error end,
+            throw({type_error, Tag, Op, P, ResTy})
     end.
+
+%% What types should be pushed into the arguments if checking an operator
+%% application against a given type.
+
+%% any() is always fine
+arith_op_arg_types(_Op, Ty = {type, _, any, []}) ->
+    {Ty, Ty};
+
+%% '/' can't produce an integer type
+arith_op_arg_types('/', Ty) when ?is_int_type(Ty) ->
+    false;
+
+%% integer() is closed under all operators except '/'
+arith_op_arg_types(_, Ty = {type, _, integer, []}) ->
+    {Ty, Ty};
+
+%% float() is closed under non-integer-exclusive operations and accepts
+%% any number() for '/'.
+%% Some precision lost: if at least one argument is a float() the other can
+%% be integer().
+arith_op_arg_types(Op, Ty = {type, _, float, []}) ->
+    case Op of
+        _ when Op == '+'; Op == '*'; Op == '-' ->
+            {Ty, Ty};
```

Comment:
Yes, but it's not clear how to do it in a nice way. We could tuple the arguments and check them against a union of tuple types, but I'm not sure how well that would work (i.e. what kind of error messages you'd get).

Reply:
Yes, or use backtracking on the outside..

The important thing is to not give errors for correct code. Better allow to much and miss some type errors. The easy way would be to push {type(number), type(number)} here I guess.

Reply:
This would be wrong though. `integer() + integer()` is not a `float()`.

Also, I don't agree that not giving errors for correct code is that important. Any static type system will reject some correct (not failing at runtime) code. The important thing is to design it to make it easy for the user to understand and predict when code will be rejected.

In this case getting a type error on `integer() + float() :: float()` might be an unhappy surprise, so it would be nice to make it work. On the other hand, no one would be surprised to discover that the type checker doesn't accept `(2..3) * (2..3) :: 4 | 6 | 9` even though it's very much correct.

Reply:
Nice that we disagree on something. :-) This is general topic. I'd like to have @josefs opinion on this.

Adding a type checker to a dynamic language with existing code bases is different than introducing a static language with a type checker in place from the beginning IMO. There's a risk that errors for correct code will not be appreciated by the community.

Reply:
I'm not sure we *do* disagree :-), but let me make an attempt at locating the disagreement more precisely.

My point is that it's impossible (halting problem etc etc) to accept all "correct" code without also accepting some incorrect code (and I don't think you're arguing that we should drop soundness). The only question then, is where to draw the line. My guess is that this is where we disagree.

My position is that it's better to find an easy to predict boundary between accepted and non-accepted programs at the price of rejecting some programs that we could conceivably accept, and your position(?) is that if we can accept a program without too much implementation effort or sever performance problems, we should, even if that makes it harder to predict when a program will type check.

In any case I think we both agree that `integer() + float() :: float()` falls within the predictability horizon (technical term I just made up).

Reply:
I'm tempted by the solution suggested by @zuiderkwast but @UlfNorell, you make a good point about `integer() + integer()` not being a `float()`. We want the type system to be sound (or, more correctly, have the graduality property. But that's such a mouthfull).
This comes down to whether we run the risk of accepting a program which crashes with a type error at runtime if we allow `integer() + integer() :: float()`. I can't think of such a program (except for programs that don't have exhaustive pattern matching. But I'd like to put them aside for now, as the Gradualizer doesn't, and probably never will, make any claims about guaranteeing exhaustiveness). Is it possible to make such a crashing program?

Reply:
```
1> float_to_list(1).
** exception error: bad argument
     in function  float_to_list/1
        called as float_to_list(1)
```

Reply:
@josefs ah, those need to be added to the documentation. It would be very good if we could make them inlineable as well, but for that I suppose it needs to be delivered as a header file instead? Are you OK with me converting `ty` to a header file, add `-compile(inline, [...])`, and prepend all the functions with `ty_`?

Reply:
@Zalastax would you like to inline them (`ty` calls) to eliminate performance penalty of a dummy function call? (I was thinking about optionally addressing this by a parse transform, but inlining is an even better idea - although I guess the overhead is very small anyway)

Reply:
@gomoripeti, yes exactly! The overhead might be small but I know from experience that it has to be absolutely free for there to not be any push back from potential critics.

Reply:
@Zalastax, just to be clear, the `ty` library doesn't solve your problem, unless I misunderstood you. I encourage you to first think about what such a library would look like. We can think about runtime performance later.

Reply:
@josefs I think it solves my problem actually 😃. Of course I'd eventually like to have a syntax for working directly on the type level (à la TypeScript) but abusing id-functions is a solution I accept for now since it's what fits within our limits. All terms can (as you said) have their type modified using this style: `int(any(T)) % T is now an int`. It's unfortunate that every type assertion will need their own function, but hey, it works. If macros were a bit stronger they could have helped us define these functions (perhaps even as they are used)...

What did you think I had in mind? I might be missing something crucial so please share!

Reply:
@Zalastax, I thought you had something in mind where you wouldn't have to spell out the call to the `any()` function. For instance, a function like this:

```
-spec assert_int(any()) -> integer().
assert_int(X) -> X.
```

But as you say, it's easy to emulate type assertion with the `ty` library, as you show. Just somewhat less convenient. I suppose we could add assertion functions to `ty` as well. We'd have to be careful about naming though. A function with the name `int()`, does it require its argument to be of type `integer()` or is it an assertion-style function? Well, perhaps it is as simple as I suggested above and prefix assertion-style function names with "assert_". What do you think?

Reply:
I got an idea. I think it is possible to write a parse transform which allows us to write type annotations on this form: `'::'(42, pos_integer())`.

The parse transform rewrites the 2nd argument in calls to '::' into a tuple representation of the type. It also inserts a function definition `'::'(Value, _Type) -> Value.` in the module, which can be inlined. In the typechecker we then trap calls to '::' and voila, we have the type.

Example: A call to `'::'([42], nonempty_list(pos_integer()))` is transformed into `'::'([42], {type, 0, nonempty_list, [{type, 0, pos_integer, []}]}`.

Would it work? Would it be good?

Reply:
That's awesome @zuiderkwast! That would be much nicer than my `ty` library.

Reply:
I don't think I could get a parse transform into my organization, that's why I've focused on other means. Regarding your suggestion: I initially didn't get all the nuances but after thinking about it for a while I think it's a very clean way to approach this!

Reply:
It would also be possible to transform calls to '::' into calls to the library functions in `ty` or whatever equivalent library we'll end up with.

Reply:
If we can come up with an annotation syntax that doesn't do anything, the code can even work without a transform. `begin "nonempty_list(pos_integer())", [42] end` for example (i.e. a block with exactly two expressions where the first one is a string literal).

Reply:
I mean, without even a dependency to a `ty` library or to gradualizer.

Reply:
@josefs that's a great idea! We can then have a special case for the `::`-function and calculate its spec from the literal in the second argument 😃